fix(hybrid search results): fixed hybrid search by removing redundant…

rajeshj11 opened the pull request:
/claim #730 
/fixes #730 

… slicing before filtering

results were sliced and then filtered, causing incorrect document return and an upper-bounded count value. Now, filtering is done first to ensure accurate total count and correct document retrieval.

demo video:
https://www.loom.com/share/63cc6cfee37c434596d79750977d5dff

rajeshj11 commented:
> Hi, can you please add some tests? Thank you

In project, we already has test cases.
packages/orama/tests/search.hybrid.test.ts

rajeshj11 commented:
> Hi, can you please add some tests? Thank you

Added the test cases. can you please check now

rajeshj11 commented:
@micheleriva are there any other changes required from my side to proceed further?  let me know if needed.

rajeshj11 commented:
@micheleriva Hey, could you please review the PR changes? if any changes are needed please let me know

nicolastoulemont commented on packages/orama/tests/search.hybrid.test.ts:
```diff
@@ -146,10 +146,13 @@ t.test('hybrid search', async (t) => {
-    t.equal(page3.count, 20)
-
+    //ideally this has to show 2. not sure why it is like this
+    // t.equal(page1.count, 20)
```
The count is supposed to be equal to the total number of documents matching the query, not the paginated amount of documents returned

rajeshj11 replied:
what about the second test case?. that count should also be 20

rajeshj11 replied:
I am confused. in the below example. the count is 2 even though the total count is 20.
https://codesandbox.io/p/sandbox/objective-sun-kzs96j?file=%2Fsrc%2Findex.ts

![Uploading Screenshot from 2024-06-10 15-06-10.png…]()

nicolastoulemont replied:
My bad, I wasn't clear enough by:
>  total number of documents matching the query
I meant the total number of documents matching the query regardless of the offset / limit

rajeshj11 commented:
@nicolastoulemont Hey, reverted the test cases to their original values. can you please check now

nicolastoulemont commented on packages/orama/tests/search.hybrid.test.ts:
```diff
@@ -198,3 +197,107 @@ t.test('hybrid search', async (t) => {
   })
 })
+
+t.test('should fix the issue realted #730', async (t) => {
```
Please write a better test name, describing in the code the intended behavior

nicolastoulemont commented on packages/orama/tests/search.hybrid.test.ts:
```diff
@@ -198,3 +197,107 @@ t.test('hybrid search', async (t) => {
+    limit: 2,
+    offset: 2,  
+  });
+  const page4 = await search(db, {
```
Not sure the page4 is necessary here tbh

nicolastoulemont reviewed (approved):
LGTM overall, would love maintainers reviews tho

rajeshj11 commented:
> LGTM

rajeshj11 commented:
@micheleriva 
 I wanted to follow up on the latest changes, where I’ve addressed all the comments and made the requested changes. Could you please take a moment to review the updates? Your feedback would be greatly appreciated.

SaraVieira commented:
Hey!

Sorry for the delay on getting back to you on this, this looks great! Merging it now

rajeshj11 commented:
@SaraVieira Ideally Post merge bounty should be rewarded. However, I am not able to see the reward label. Is this issue from algora pbc bot?

micheleriva commented:
Hi @rajeshj11 please give us one day to process this.

rajeshj11 commented:
> Hi @rajeshj11 please give us one day to process this.

yeah sure, no rush. Just want to know where it is blocked.

rajeshj11 commented:
@micheleriva may I know the status.
